README.md
# Matcher

[![Clojars Project](https://img.shields.io/clojars/v/matcher.svg)](https://clojars.org/matcher)

Matcher is an extensible Clojure pattern matching library. It intends to be simple, extensible, powerful, and idiomatic - for example, `match` syntax is almost the same of `cond`, so no surprises here.

The important keywords are **simple** - _matcher_ depends on `core.unify`, so all variable bindings are very simple, and all matchers are just curried functions, so they are partially evaluated every time, so no magic - and **extensible** - there are _no_ "built-in matchers - everything must be or a single value or a function. No strange huge and un-debuggable macros to deconstruct matchers.

## Usage

The core of everything is on `matcher.matches` namespace. Each match clause is a value, or a function that receives a single argument (more on that later).

```clojure
(require '[matcher.matches :as m])

(defn example [some-obj]
  (m/match some-obj
    10 "some-obj is ten"
    (m/list ?a & _) (str "it is a list, and first arg is " a)
    string? "It is a string"
    _ "I don't know what some-obj is..."))
```

The namespace `matcher.matches` contains a bunch of *matcher functions*. These are special functions that will check specific rules and bind variables if matched.

```clojure
(m/match some-obj
  (m/list ?first ?second 3) (+ first second))
```

The code above will try to match a list (it will **not** match vectors or maps) that have three elements, and the third is `3`. It will bind the first element to the local variable `first` and will bind the second variable to `second`.

More examples can be found on the `examples` directory.

## Matcher Functions

Any function that receive only one argument, and return true or false, can be used as a matcher. So, `odd?`, `even?`, `map?` and such can be used as matchers, but they _will not_ bind variables nor check for exceptions. So, the code below will throw an error because `"10"` is not a number:

```clojure
(m/match "10"
  odd? "It's a wild world...")
```

To avoid exceptions and bind variables to the match functions, it is better to use a specific _matcher function_. One of these functions is `satisfies` that will wrap the predicate in a try-catch, and it'll allow us to capture the result of predicate. The code below, for instance, returns `40`, the result of the only `some` that matches (the second one):

```clojure
(m/match '(10 20 30 40 50)
  (m/satisfies #(some #{-1 -2} %) ?r) (str "Negative number found: " r)
  (m/satisfies #(some #{3 4 50 40} %) ?r) (str "Positive number found: " r))
```

Please notice that matchers are **not** special in any way - they're just Clojure functions. `satisfies`, for instance, is just a function that returns another function, wrapping the check in a try-catch block. You could, for instance, check what `satisfies` returns by just running it by hand (and quoting `?r` and other symbols):

```clojure
(m/satisfies #(some #{3 4 50 40} %) '?r) ; => returns a function
((m/satisfies #(some #{3 4 50 40} %) '?r) '(10 20 30 40 50))
; => returns ['?r 40]
```

On `matcher.matchers` namespace, there are already a bunch of matchers prepared (with examples):

* `list`, `vector`, and `coll` - matchers for collections. To correcly match, you need to either provide each individual element or use `&` to match the rest of the list. For example `(list)` will match any empty list (but not vector), `(vector 10 & _)` matches any vector that begins with `10`, and `(coll 10 20)` will match either a list or vector with two elements, `10` and `20`.
```clojure
(m/match [1 2 3 4]
  (m/list 1 2 3 4) "won't match - its not a list"
  (m/vector 1 2 3) "won't match - missing one element"
  (m/vector 1 ?a ?a 4) "won't match - second and third elements are not equal"
  (m/vector 1 2 3 4) "will match"
  (m/vector 1 _ & ?rest) "would match, but previous matcher catch the code")
```
* `matches` is a generic unifier. It'll just call `core.unify/unify`. This matcher is *not* composable: you cannot use `(matches odd?)` for example. In general, avoid this match, unless you want to match maps or map-like structures like records, for instance, and don't want the extra overhead introduced by `map` matcher.
```clojure
(m/match '(1 2 3 4)
  (m/matches [1 2 ?a ?a]) "will not match - third and fourth are not equal"
  (m/matches (m/vector 1 2 3 4)) "will not match - matches does not allow inner matchers"
  (m/matches [1 2 3 4]) "will match - matches uses unify, which does not differentiate colls")
```
A possible use would be to match maps faster:
```clojure
(m/match {:a 10 :b 20}
  (m/matches {:c ?a}) "will not match - there's no :c in map"
  (m/matches {?a 10}) "will not match - missing :b"
  (m/matches [[?a 10] & _]) "will match, because it treats all colls the same")
```

* `satisfies` wraps a predicate (like `odd?`, `even`, etc) in a try-catch, so that they will not throw exceptions if called with wrong arguments. Furthermore, it allows another parameter that will try to unify with the predicate's result. `satisfies` does not accept inner matches - so, something like `(satisfies to-int odd?)` will not work.
```clojure
(m/match "10"
  (m/satisfies even?) "will not match - it's not a number"
  ; Please, notice that parseInt need to be in anon function because Clojure
  ; can't identify it as a function (because it's Java interop):
  (m/satisfies #(Integer/parseInt %) ?n) (str "will match, and the number is: " n))
```

* `map` is a complicated matcher. It'll try to match keys *and* values on a map. It'll accept an even number of arguments, and will lazily check each key and value to unify it. Of course, as maps' keys are unordered, this could mean that unified keys are not previsible. For example, `(map ?a 10)`, applied to `{:a 10 :b 10 :c 10}`, could bind `a` to `:a`, `:b` or `:c`
```clojure
(m/match {:some "strange" :map "Here"}
  (m/map :some number?) "Will not match - value of :some is not a number"
  (m/map :missing _) "Will not match - there's not :missing key"
  (m/map :some) "Will not match - missing value for :some key"
  (m/map :some string?) "Will match")
```
* `regexp` asks for a regular expression, and a number of other arguments. If the regexp have capture groups, other arguments will be the capture groups, in order. So, if they match, the matcher will be successful. If they are unbound variables, they'll be bound to the regular expression's capture groups. There must be the same number of arguments that are capture groups, or none at all.
```clojure
(m/match "github.com/mauricioszabo/matcher"
  (m/regexp #".*/(.*)/(.*)" ?a ?a) "Will not match - two parts of URL are not the same"
  (m/regexp #".*/(.*)/(.*)" _ "clj") "Will not match - second part of URL is not right"
  (m/regexp #".*/(.*)/(.*)" ?a "matcher") (str "Will match, and capture " a)
  (m/regexp #".*/(.*)/(.*)") "Would match, but previous matcher already captured")
```
* `instance` simply checks if object is an instance of a specific parameter. For example, trying to match a list, `(m/instance clojure.lang.PersistentList)` will match, and `(m/instance ?c)` will not match (it doesn't makes sense to bind variables to something we could just use `(type ..some-obj..)`.
```clojure
(m/match "Foo Bar"
  (m/instance Integer) "Will not match - it's not an Integer"
  (m/instance 'String) "Will not match - it's not a class"
  (m/instance String) "Will match")
```

* `record` checks if object is a specific record, and tries to unify its arguments (in order of definition). See examples for more info.
```clojure
(defrecord RGB [r g b])
(defrecord RGBA [r g b a])
((m/record RGB) (->RGB 10 20 30))
(m/match (->RGB 10 20 30)
  (m/record RGBA _ _ _ _) "Will not match - it's not a RGBA"
  (m/record RGB ?a ?a _) "Will not match - r and g are different"
  (m/record RGB 10 ?g 30) (str "Will match and capture green: " g))
```

## Custom matchers
One of the core ideas for Matcher is to be extensible - so, it's possible to define your own matcher on it.

A matcher is simply a function that, when applied to the matching object, will return truthy or falsey. If you want to create a matcher that'll bind variables, you **need** to return a vector with exactly two elements. The "left side" is the parameters passed to test, if any; the "right side" is the matching part. This will be passed to `clojure.core.unify/unify`, and if that returns a map, it'll be a match. If the map have unbound variables (variables that starts with `?`, like `?foo`), let-variables will be created.

One other thing is that if you want matcher to recurse into results (that is, if we want Matcher to test inner matchers), then we must assure that the left-side and right-side are colls (vectors, lists, maps or sets).

For example, let's create a matcher that will ask what's the previous prime of that number, if any. First of all, remember that what the matcher will receive can be an integer, or not - so, we need to check for this too. Second, remember that the user can bind a variable on it, or can just check if the last-prime is a specific number (but `unify` will check for us, so we don't need to check for it).

So, we'll define a function named `last-prime` that will expect an argument, and it'll return another function. This other function will: (1), check if the parameter passed to it is an integer; (2), check if the number is greater than `2` (because there's no primes before `2`), and (3), return a vector with what we expect the last-prime to be, and what the last-prime really is, wrapped in colls (first in a vector, second in a list):

```clojure
(defn last-prime [number]
  (fn [possible-number]
    (when (integer? possible-number)
      [[number]
       (->> (range possible-number 2 -1)
            (remove (fn [x] (some #(-> x (rem %) zero?) (range 2 x))))
            (take 1))])))
```

Simple code, line 5 to 7 calculates the prime, and assigns it to the second element of the vector. The first argument of the vector is just the number we expect. If the two numbers are different, `unify` will take care for us. If the first number is an unbound variable, it'll bind the variable for us. And, if the first argument is another matcher, then **Matcher** will take care for us. So, we can just use the matcher:

```clojure
(m/match 10
  (last-prime 9) "Will not match - 9 is not even prime!"
  (last-prime 5) "Will not match - it is not 5"
  (last-prime even?) "Will not match - it's odd"
  (last-prime odd?) "Will match - it's odd")
```

If we remove the "vector" over the number (line 4) and change `(take 1)` to `first`, we'll see that the matcher still works, but it'll not recurse into inner matches (like `odd?` or `even?` - these lines would always return false, and the code above would only check for unbound vars or numbers.

### Custom and composite matchers
There are matchers (like `and` and `or`) that need to know the result from previous matchers to be able to decide if they match or not. Let's see how to create a `not` matcher, one that negates the result of current matcher (please note: because of the way matchers work, this is _really_ a bad idea - unbound variables will all be bound to `nil`, for instance, but this is only an example).

If we look at `last-prime`, for example, we can see that only running the matcher's function is not sufficient to check if there was a match or not. So, we need to import `matcher.utils.match`, a bunch of utility functions that recursively checks for matches:

```clojure
(require '[matcher.utils.match :as utils])
(utils/match-and-unify 10 (last-prime 9)) ; => nil
(utils/match-and-unify 10 (last-prime 7)) ; => {}
(utils/match-and-unify 10 (last-prime '?n)) ; => {'?n 7}
```

So, a `not` matcher would be just:

```clojure
(require '[matcher.utils.match :as utils])
(defn not-matcher [matcher]
  (fn [obj]
    (when (not (utils/match-and-unify obj matcher))
      true)))

(m/match [1 2 3]
  (not-matcher (m/vector 1 2 3)) "Will not match"
  (not-matcher (m/list :foo :bar)) "Match!")
```

For more information, look at how `or` and `and` matchers are implemented.

## License

Copyright © 2016 Maurício Szabo

Distributed under the Eclipse Public License either version 1.0 or any later version.


CHANGELOG.md
# Change Log

## 0.0.6
- Corrected record matcher to accept no arguments

## 0.0.5
### Changed
- Added match macro
- Added matchers


test/matcher/matchers_test.clj
(ns matcher.matchers-test
  (:require [midje.sweet :refer :all]
            [matcher.matchers :as m]
            [clojure.core.unify :as u]))

(defn match-and-unify [ & {:as map}]
  (fn [obj]
    (= map (apply u/unify obj))))

(defn dont-match [obj]
  (or (nil? obj) (nil? (apply u/unify obj))))

(defn match [obj]
  (= {} (apply u/unify obj)))

(def test-list (list 1 2 3))
(fact "Matching against lists"
  ((m/list 1 2 3) [1 2 3]) => dont-match
  ((m/list 1 2 3) test-list) => match
  ((m/list 1 '?foo 3) test-list) => (match-and-unify '?foo 2)
  ((m/list 1 '& '?rest) test-list) => (match-and-unify '?rest '(2 3)))

(def test-vec [1 2 3])
(fact "Matching against vectors"
  ((m/vector 1 2 3) '(1 2 3)) => nil
  ((m/vector 1 2 3) test-vec) => [[1 2 3] [1 2 3]]
  ((m/vector 1 '?foo 3) test-vec) => [[1 '?foo 3] [1 2 3]]
  ((m/vector 1 '& '?rest) test-vec) => [[1 '?rest] [1 '(2 3)]])

(fact "Matching against any collection"
  ((m/coll 1 2 3) '(1 2 3)) => match
  ((m/coll 1 '?foo 3) test-vec) => (match-and-unify '?foo 2))

(def test-map {:a 10, :b 20})
(fact "Matching using unifier only"
  ((m/matches {:a '?a, '?b 20}) test-map) => [{:a '?a, '?b 20} {:a 10 :b 20}]
  ((m/matches {:a '?a, '?b 20}) test-vec) => [{:a '?a, '?b 20} [1 2 3]]
  ((m/matches [1 2 3]) test-vec) => [[1 2 3] [1 2 3]])

(facts "Predicate matching with core functions"
  ((m/satisfies even?) 10) => match
  ((m/satisfies even? '?result) 10) => (match-and-unify '?result true)
  ((m/satisfies even? '?result) "FOO") => dont-match)

(facts "Matching against maps"
  (fact "simple matches"
    ((m/map :a '?a) test-vec) => dont-match
    ((m/map) test-map) => match
    ((m/map :a '?a) test-map) => (match-and-unify '?a 10)
    ((m/map '?a 20) test-map) => (match-and-unify '?a :b)
    ((m/map '?a '?b) test-map) => (match-and-unify '?a :a '?b 10)
    ((m/map '?a 20 '?b 10) test-map) => (match-and-unify '?a :b '?b :a))

  (fact "composite matches"
    ((m/map :a (m/list)) test-map) => dont-match
    ((m/map :a 10 :b (m/list)) test-map) => dont-match
    ((m/map '?a (m/list '& '_)) {:a 10, :b 20, :c '(1 2 3)}) => (match-and-unify '?a :c)))

(fact "Matching strings"
  (fact "matches regexp"
    ((m/regexp #"(.*)@(.*)" "bar" '?b) "foo@bar") => dont-match
    ((m/regexp #"(.*)@(.*)") "foo@bar") => match
    ((m/regexp #"(.*)@(.*)" '?a '?b) "foo@bar") => (match-and-unify '?a "foo" '?b "bar")))

(fact "Matching instances of some element"
  ((m/instance clojure.lang.PersistentList) '(1 2) ) => []
  ((m/instance clojure.lang.PersistentList) [1 2] ) => dont-match
  ((m/instance 10) [1 2] ) => dont-match)

(defrecord Example [a b c])
(defrecord Example2 [a b c])
(def test-rec (->Example 10 20 30))
(fact "Matching records"
  ((m/record Example '?a '?b) test-rec) => dont-match
  ((m/record Example2 '?a '?b '?c) test-rec) => dont-match
  ((m/record Example '?a '?b '?c) test-rec) => (match-and-unify '?a 10 '?b 20 '?c 30)
  ((m/record Example 10 '?b '?c) test-rec) => (match-and-unify '?b 20 '?c 30)
  ((m/record Example) test-rec) => match

  ((m/record 'Example3 '?a '?b '?c) test-rec) => dont-match
  ((m/record java.lang.Byte '?a '?b '?c) test-rec) => dont-match
  ((m/record matcher.matchers_test.Example '?a '?b '?c) test-rec)
  => (match-and-unify '?a 10 '?b 20 '?c 30))

(facts "boolean matches"
  (fact "or applies first rule that matches"
    ((m/or (m/list 1 '?l 3) (m/list 1 '?v 3)) test-vec) => dont-match
    ((m/or (m/list 1 '?l 3) (m/vector 1 '?v 3)) test-vec) => (match-and-unify '?v 2)
    ((m/or (m/vector '?first 2 3) (m/vector 1 '?v 3)) test-vec) => (match-and-unify '?first 1))

  (fact "and applies only if all rules apply"
    ((m/and (m/vector 1 '?l 3) (m/list 1 1 3)) test-vec) => dont-match
    ((m/and (m/vector 1 '?v 3) (m/vector 1 2 '?v)) test-vec) => dont-match
    ((m/and (m/vector 1 '?v1 '?l) (m/vector 1 '?v2 '?l)) test-vec)
    => (match-and-unify '?v1 2 '?v2 2 '?l 3)))
    ; ((m/and (m/vector '?first 2 3) (m/vector 1 '?v 3)) test-vec) => (match-and-unify '?first 1)))

(facts "Using match macro"
  (fact "matching literal values"
    (m/match 10
      10 :foo
      20 :bar) => :foo)

  (fact "matching predicate fns"
    (m/match 20
      odd? :is-odd
      even? :is-even) => :is-even)

  (fact "matching without vars"
    (m/match test-list
      (m/list 1 2 1) :foo
      (m/list 1 2 3) :bar) => :bar

    (m/match '(1 2 (2 3) 4)
      (m/list 1 2 '(2 2) 4) :foo
      (m/list 1 2 (m/list 2 3) 4) :bar) => :bar)

  (fact "matching with vars"
    (m/match test-list
      (m/list ?a 2 ?a) :foo
      (m/list ?a 2 ?b) (+ a b)) => 4

    (m/match '(1 2 (2 3) 4)
      (m/list ?a 2 (m/list ?a ?a) 4) :foo
      (m/list 1 ?a (m/list ?a ?b) ?c) (+ a b c)) => 9)

  (fact "applies inner params and anon functions"
    (m/match '(30 40 50)
      (m/matches {:a ?b}) :foo
      (m/satisfies #(some #{-1 -2} %) ?r) r
      (m/satisfies #(some #{3 4 50 40} %) ?r) r) => 40))


test/matchbox/matchers_test.clj
(ns matchbox.matchers-test
  (:require [midje.sweet :refer :all]
            [matchbox.matchers :as m]
            [clojure.core.unify :as u]))

(defn match-and-unify [ & {:as map}]
  (fn [obj]
    (= map (apply u/unify obj))))

(defn dont-match [obj]
  (or (nil? obj) (nil? (apply u/unify obj))))

(defn match [obj]
  (= {} (apply u/unify obj)))

(def test-list (list 1 2 3))
(fact "Matching against lists"
  ((m/list 1 2 3) [1 2 3]) => dont-match
  ((m/list 1 2 3) test-list) => match
  ((m/list 1 '?foo 3) test-list) => (match-and-unify '?foo 2)
  ((m/list 1 '& '?rest) test-list) => (match-and-unify '?rest '(2 3)))

(def test-vec [1 2 3])
(fact "Matching against vectors"
  ((m/vector 1 2 3) '(1 2 3)) => nil
  ((m/vector 1 2 3) test-vec) => [[1 2 3] [1 2 3]]
  ((m/vector 1 '?foo 3) test-vec) => [[1 '?foo 3] [1 2 3]]
  ((m/vector 1 '& '?rest) test-vec) => [[1 '?rest] [1 '(2 3)]])

(fact "Matching against any collection"
  ((m/coll 1 2 3) '(1 2 3)) => match
  ((m/coll 1 '?foo 3) test-vec) => (match-and-unify '?foo 2))

(def test-map {:a 10, :b 20})
(fact "Matching using unifier only"
  ((m/matches {:a '?a, '?b 20}) test-map) => [{:a '?a, '?b 20} {:a 10 :b 20}]
  ((m/matches {:a '?a, '?b 20}) test-vec) => [{:a '?a, '?b 20} [1 2 3]]
  ((m/matches [1 2 3]) test-vec) => [[1 2 3] [1 2 3]])

(facts "Predicate matching with core functions"
  ((m/satisfies even?) 10) => match
  ((m/satisfies even? '?result) 10) => (match-and-unify '?result true)
  ((m/satisfies even? '?result) "FOO") => dont-match)

(facts "Matching against maps"
  (fact "simple matches"
    ((m/map :a '?a) test-vec) => dont-match
    ((m/map) test-map) => match
    ((m/map :a '?a) test-map) => (match-and-unify '?a 10)
    ((m/map '?a 20) test-map) => (match-and-unify '?a :b)
    ((m/map '?a '?b) test-map) => (match-and-unify '?a :a '?b 10)
    ((m/map '?a 20 '?b 10) test-map) => (match-and-unify '?a :b '?b :a))

  (fact "composite matches"
    ((m/map :a (m/list)) test-map) => dont-match
    ((m/map :a 10 :b (m/list)) test-map) => dont-match
    ((m/map '?a (m/list '& '_)) {:a 10, :b 20, :c '(1 2 3)}) => (match-and-unify '?a :c)))

(fact "Matching strings"
  (fact "matches regexp"
    ((m/regexp #"(.*)@(.*)" "bar" '?b) "foo@bar") => dont-match
    ((m/regexp #"(.*)@(.*)") "foo@bar") => match
    ((m/regexp #"(.*)@(.*)" '?a '?b) "foo@bar") => (match-and-unify '?a "foo" '?b "bar")))

(fact "Matching instances of some element"
  ((m/instance clojure.lang.PersistentList) '(1 2) ) => []
  ((m/instance clojure.lang.PersistentList) [1 2] ) => dont-match
  ((m/instance 10) [1 2] ) => dont-match)

(defrecord Example [a b c])
(defrecord Example2 [a b c])
(def test-rec (->Example 10 20 30))
(fact "Matching records"
  ((m/record Example '?a '?b) test-rec) => dont-match
  ((m/record Example2 '?a '?b '?c) test-rec) => dont-match
  ((m/record Example '?a '?b '?c) test-rec) => (match-and-unify '?a 10 '?b 20 '?c 30)

  ((m/record 'Example3 '?a '?b '?c) test-rec) => dont-match
  ((m/record java.lang.Byte '?a '?b '?c) test-rec) => dont-match
  ((m/record matchbox.matchers_test.Example '?a '?b '?c) test-rec)
  => (match-and-unify '?a 10 '?b 20 '?c 30))

(facts "boolean matches"
  (fact "or applies first rule that matches"
    ((m/or (m/list 1 '?l 3) (m/list 1 '?v 3)) test-vec) => dont-match
    ((m/or (m/list 1 '?l 3) (m/vector 1 '?v 3)) test-vec) => (match-and-unify '?v 2)
    ((m/or (m/vector '?first 2 3) (m/vector 1 '?v 3)) test-vec) => (match-and-unify '?first 1))

  (fact "and applies only if all rules apply"
    ((m/and (m/vector 1 '?l 3) (m/list 1 1 3)) test-vec) => dont-match
    ((m/and (m/vector 1 '?v 3) (m/vector 1 2 '?v)) test-vec) => dont-match
    ((m/and (m/vector 1 '?v1 '?l) (m/vector 1 '?v2 '?l)) test-vec)
    => (match-and-unify '?v1 2 '?v2 2 '?l 3)))
    ; ((m/and (m/vector '?first 2 3) (m/vector 1 '?v 3)) test-vec) => (match-and-unify '?first 1)))

(facts "Using match macro"
  (fact "matching literal values"
    (m/match 10
      10 :foo
      20 :bar) => :foo)

  (fact "matching predicate fns"
    (m/match 20
      odd? :is-odd
      even? :is-even) => :is-even)

  (fact "matching without vars"
    (m/match test-list
      (m/list 1 2 1) :foo
      (m/list 1 2 3) :bar) => :bar

    (m/match '(1 2 (2 3) 4)
      (m/list 1 2 '(2 2) 4) :foo
      (m/list 1 2 (m/list 2 3) 4) :bar) => :bar)

  (fact "matching with vars"
    (m/match test-list
      (m/list ?a 2 ?a) :foo
      (m/list ?a 2 ?b) (+ a b)) => 4

    (m/match '(1 2 (2 3) 4)
      (m/list ?a 2 (m/list ?a ?a) 4) :foo
      (m/list 1 ?a (m/list ?a ?b) ?c) (+ a b c)) => 9)

  (fact "applies inner params and anon functions"
    (m/match '(30 40 50)
      (m/matches {:a ?b}) :foo
      (m/satisfies #(some #{-1 -2} %) ?r) r
      (m/satisfies #(some #{3 4 50 40} %) ?r) r) => 40))


test/matcher/utils/match_test.clj
(ns matcher.utils.match-test
  (:require [midje.sweet :refer :all]
            [matcher.utils.match :as utils]
            [matcher.matchers :as m]
            [clojure.core.unify :as u]))

(def foo 10)
(fact "parses args correctly from matchers"
  (utils/parse-args '(m/list ?arg _ 10 foo)) => '(m/list '?arg '_ 10 foo)
  (utils/parse-args '(m/list _ & _foo)) => '(m/list '_ '& _foo)
  (utils/parse-args '(m/list ?arg _ (m/list ?b) foo)) => '(m/list '?arg '_ (m/list '?b) foo)
  (utils/parse-args '(m/list [?b _ #{_ ?a}])) => '(m/list ['?b '_ #{'_ '?a}]))

(def test-list (list 1 2 3))
(facts "parses match args correcly"
  (utils/apply-match test-list (m/list 1 2 3)) => [[1 2 3] [1 2 3]]
  (utils/apply-match test-list (m/list '?a 2 3)) => [['?a 2 3] [1 2 3]]
  (utils/apply-match test-list (m/list 1 2)) => nil

  (fact "parses single value matchers"
    (utils/apply-match nil nil?) => [[] []]
    (utils/apply-match nil nil) => [[nil] [nil]]))

(def inner-list (list 1 (list 2 3) 4))
(facts "parses inner match args correctly"
  (utils/apply-match 10 20)
  => [[20] [10]]
  (utils/apply-match inner-list (m/list 1 (m/list 2 3) 4))
  => [[1 [2 3] 4] [1 [2 3] 4]]
  (utils/apply-match inner-list (m/list '?a (m/list '?a '?b) 4))
  => [['?a ['?a '?b] 4] [1 [2 3] 4]]
  (utils/apply-match test-list (m/list 1 (m/list 2 3) 4)) => nil)

(facts "matching patterns"
  (background
    (gensym) => 'foo)

  (fact "unwraps code if there are no sexps on match side"
    (utils/wrap-let 10 10 :foo :bar)
    => `(if-let [~'foo (utils/match-and-unify 10 10)]
         :foo
         :bar))

  (fact "unwraps code if matcher is a set"
    (utils/wrap-let 10 #{1 10} :foo :bar)
    => `(if-let [~'foo (utils/match-and-unify 10 #{1 10})]
         :foo
         :bar))

  (fact "adds an let if there are unbound vars"
    (utils/wrap-let 'test-list '(m/list 10 (m/list 20) 30) ':foo ':bar)
    => `(if-let [~'foo (utils/match-and-unify ~'test-list (~'m/list 10
                                                                    (~'m/list 20)
                                                                    30))]
         :foo
         :bar)

    (utils/wrap-let 'test-list '(m/list ?a (m/list ?a ?b) _) '(+ a b) ':bar)
    => `(if-let [~'foo (utils/match-and-unify ~'test-list (~'m/list ~''?a
                                                                    (~'m/list ~''?a ~''?b)
                                                                    ~''_))]
          (let [~'a (~''?a ~'foo)
                ~'b (~''?b ~'foo)]
            (~'+ ~'a ~'b))
          :bar)

    (utils/wrap-let 'test-list '(m/match {:a ?a :b ?b}) '(+ a b) ':bar)
    => `(if-let [~'foo (utils/match-and-unify ~'test-list (~'m/match {:a ~''?a
                                                                      :b ~''?b}))]
          (let [~'a (~''?a ~'foo)
                ~'b (~''?b ~'foo)]
            (~'+ ~'a ~'b))
          :bar)))

(facts "pattern matching without macros"
  (fact "expands macro in a sane way"
    (utils/match* 'test-list
      '(m/list 10) :foo)
    => `(if-let [f ..first-clause..]
          :foo
          (throw (IllegalArgumentException. "No match")))
    (provided
      (utils/wrap-let 'test-list '(m/list 10) :foo `(throw (IllegalArgumentException. "No match")))
      => `(if-let [f ..first-clause..] :foo (throw (IllegalArgumentException. "No match")))))

  (fact "expands macro with two matches"
    (utils/match* 'test-list
      '(m/list 10) :foo
      '(m/list 20) :bar)
    => `(if ..c1.. ..10.. (if ..c2.. ..20.. ..throw..))
    (provided
      (utils/wrap-let 'test-list '(m/list 20) :bar `(throw (IllegalArgumentException. "No match")))
      => `(if ..c2.. ..20.. ..throw..)
      (utils/wrap-let 'test-list '(m/list 10) :foo `(if ..c2.. ..20.. ..throw..))
      => `(if ..c1.. ..10.. (if ..c2.. ..20.. ..throw..))))

  (fact "adds catch-all"
    (utils/match* 'test-list
      '(m/list 10) :foo
      '_ :bar)
    => `(if ..c1.. ..10.. :bar)
    (provided
      (utils/wrap-let 'test-list '(m/list 10) :foo :bar)
      => `(if ..c1.. ..10.. :bar))))


examples/matching.clj
(require '[matcher.matchers :as m])

(def some-list '(1 2 40 50))

; Feel free to vary any parameter below...
(println (m/match some-list
           (m/list odd? even? odd? even?) "Alternate numbers"
           (m/list odd? odd? odd? odd?) "All odd"
           (m/list even? even? even? even?) "All Even"
           (m/list _ _ even? even?) "Last two are even"
           _ "Didn't find a match"))

(println (m/match some-list
           (m/list ?f ?s odd? even?) (str "Alternate numbers, sum of first two: " (+ f s))
           (m/list odd? odd? & ?rest) (str "First two are odd, rest of the list: " (vec rest))
           (m/vector ?f ?s even? even?) (str "Vector, last two are even, sum of first two: " (+ f s))
           (m/list ?f ?s even? even?) (str "List, last two are even, sum of first two: " (+ f s))
           _ "Didn't find a match"))

; String matchers
(m/match "foo@bar.com"
  (m/regexp #"(.*)@(.*)" ?user ?domain) (println user "AT" domain))

(m/match "foo@bar.com"
  (m/regexp #"(.*)@(.*)" ?user) (println user)
  (m/regexp #"(.*)@(.*)") (println "MATCHES!"))

; Predicate matchers
(println (m/match 20
           odd? "It's odd"
           even? "It's even"))

(println (m/match "20"
           (m/satisfies odd?) "It's odd"
           (m/satisfies even?) "It's even"
           _ "It's not a number"))

; PLEASE note that you probably want to extract the matcher below to a function.
; Matchers can be very sensitive about inner functions and such, and we don't expect
; to use "filter", "some", and other complex operations inside it.
(println (m/match some-list
           (m/satisfies #(some #{-1 -2} %) ?r) (str "Negative number found: " r)
           (m/satisfies #(some #{3 4 50 40} %) ?r) (str "Positive number found: " r)))

; A pratical example
(def some-ast
  '(+ (* 10 2) (* 20 (+ 30 20))))
(defn calculate-result [ast]
  (m/match ast
    (m/list '+ & ?rest) (reduce #(+ %1 (calculate-result %2)) 0 rest)
    (m/list '* & ?rest) (reduce #(+ %1 (calculate-result %2)) 0 rest)
    number? ast))
(println (calculate-result some-ast))

; Nested matchers
(println (m/match some-ast
           (m/list _ list? (m/list _ ?number (m/list _ ?number ?other)))
           (str "Will not match - number must be the same in both places: " (+ number other))

           (m/list _ list? (m/list _ ?number (m/list _ ?other ?number)))
           (str "Will match: " (+ number other))))

; Boolean matches
(def some-vector [3 2 3 4])
(println (m/match some-vector
           (m/or (m/vector ?a ?b) (m/list _ _ ?a ?b)) "Will not match"
           (m/or (m/vector ?a ?b) (m/vector _ _ ?a ?b)) (+ a b)))

(println (m/match some-vector
           (m/and (m/vector ?a _ _ ?b) (m/vector _ ?a _ ?b)) "Will not match - ?a cannot be 3 and 2"
           (m/and (m/vector ?a _ _ ?b) (m/vector _ _ ?a ?b)) (+ a b)))

; Records (Binary tree example)
(defrecord Tree [left value right])
(defn insert [tree new-value]
  (m/match tree
    nil (->Tree nil new-value nil)
    (m/record Tree ?left ?value ?right) (if (< new-value value)
                                          (->Tree (insert left new-value) value right)
                                          (->Tree left value (insert right new-value)))))

(defn tree->list [tree]
  (m/match tree
    nil nil
    (m/record Tree ?left ?value ?right) (concat (tree->list left) (cons value (tree->list right)))))

(-> nil
    (insert 10)
    (insert 25)
    (insert 1)
    (insert 4)
    (insert 40)
    (insert 12)
    (insert 21)
    tree->list)
